3_harmonize/src/clean_wqp_data.R

```diff
@@ -3,29 +3,171 @@
+    flag_duplicates(., duplicate_definition) %>%
+    {if(remove_duplicated_rows){
+      remove_duplicates(., duplicate_definition)
+    } else {.}
+    }
+  
+  # Inform the user what we found for duplicated rows
+  if(remove_duplicated_rows){
+    message(sprintf(paste0("Removed %s duplicated records."), 
+                    nrow(wqp_data) - nrow(wqp_data_clean)))
+  }
   
   return(wqp_data_clean)
   
+}
+
+
+#' @title Flag missing results
+#' 
+#' @description 
+#' Function to flag true missing results, i.e. when the result measure value 
+#' and detection limit value are both NA, when "not reported" is found in the
+#' column "ResultDetectionConditionText", or when any of the strings from
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a data record. Must contain the columns
+#' "DetectionQuantitationLimitMeasure.MeasureValue", "ResultMeasureValue", 
+#' "ResultDetectionConditionText", and "ResultCommentText".
+#' @param commenttext_missing character string(s) indicating which strings from
+#' the WQP column "ResultCommentText" correspond with missing result values.
+#' 
+#' @returns
+#' Returns a data frame containing data downloaded from the Water Quality Portal,
+#' where each row represents a data record. New columns appended to the original
+#' data frame include flags for missing results. 
+#' 
+flag_missing_results <- function(wqp_data, commenttext_missing){
+  
+  wqp_data_out <- wqp_data %>%
+    mutate(flag_missing_result = case_when(
+      is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ~ TRUE,
+      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) ~ TRUE,
+      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE) ~ TRUE,
+      TRUE ~ FALSE
+    ))
+  
+  return(wqp_data_out)
+  
+}
+
+
+
+#' @title Flag duplicated records
+#' 
+#' @description 
+#' Function to flag duplicated rows based on a user-supplied definition
+#' of a duplicate record. 
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a data record.
+#' @param duplicate_definition character string(s) indicating which columns are
+#' used to identify a duplicate record. Duplicate records are defined as those 
+#' that share the same value for each column within `duplicate_definition`.
+#'
+#' @returns 
+#' Returns a data frame containing data downloaded from the Water Quality Portal,
+#' where each row represents a data record. New columns appended to the original
+#' data frame include flags for duplicated records. 
+#' 
+flag_duplicates <- function(wqp_data, duplicate_definition){
+  
+  # Flag duplicate records using the `duplicate_definition`
+  wqp_data_out <- wqp_data %>%
+    group_by(across(all_of(duplicate_definition))) %>% 
+    mutate(n_duplicated = n(),
+           flag_duplicated_row = if_else(n_duplicated > 1, TRUE, FALSE)) %>% 
```

Comment:
I think this could just be written as `n_duplicated > 1` because that will already return a `TRUE` or a `FALSE` and you don't need the `if_else()`

Reply:
Oh, nice! I don't return logical within `mutate` much I guess, but I like this implementation. Added in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/74/commits/9f5213fbfaf6c625c28c61ab47d589f6abe567ca.